File: backend-server/src/controllers/student.controller.js
"use strict";

const { STATUS } = require("../config/constants");
const studentService = require("../services/student.service");

exports.createStudent = async (req, res) => {
    const { name, phoneNumber } = req.body;
    const schoolId = req.schoolId;
    if (!name || typeof name !== "string" || name.trim().length === 0 || !phoneNumber) {
        return res.status(STATUS.BAD_REQUEST).json({ message: "name and phoneNumber are required" });
    }
    try {
        const student = await studentService.createStudent(name.trim(), phoneNumber, schoolId);
        return res.status(STATUS.CREATED).json(student);
    } catch (error) {
        if (error.code === 11000) {
            return res.status(STATUS.CONFLICT).json({ message: "Phone number already in use in this school" });
        }
        return res.status(STATUS.INTERNAL_ERROR).json({ message: "Internal server error" });
    }
};

exports.getStudents = async (req, res) => {
    try {
        const students = await studentService.getStudentsBySchoolId(req.schoolId);
        return res.status(STATUS.OK).json(students);
    } catch (error) {
        return res.status(STATUS.INTERNAL_ERROR).json({ message: "Internal server error" });
    }
};

exports.updateStudent = async (req, res) => {
    const { studentId } = req.params;
    const { name, phoneNumber } = req.body;
    if (!name && !phoneNumber) {
        return res.status(STATUS.BAD_REQUEST).json({ message: "name or phoneNumber is required" });
    }
    const updates = {};
    if (name) updates.name = name.trim();
    if (phoneNumber) updates.phoneNumber = phoneNumber;
    try {
        const student = await studentService.updateStudent(studentId, req.schoolId, updates);
        return res.status(STATUS.OK).json(student);
    } catch (error) {
        if (error.status === STATUS.NOT_FOUND) {
            return res.status(STATUS.NOT_FOUND).json({ message: error.message });
        }
        if (error.code === 11000) {
            return res.status(STATUS.CONFLICT).json({ message: "Phone number already in use in this school" });
        }
        return res.status(STATUS.INTERNAL_ERROR).json({ message: "Internal server error" });
    }
};

exports.deleteStudent = async (req, res) => {
    const { studentId } = req.params;
    try {
        await studentService.deleteStudent(studentId, req.schoolId);
        return res.status(STATUS.OK).json({ message: "Student deleted successfully" });
    } catch (error) {
        if (error.status === STATUS.NOT_FOUND) {
            return res.status(STATUS.NOT_FOUND).json({ message: error.message });
        }
        return res.status(STATUS.INTERNAL_ERROR).json({ message: "Internal server error" });
    }
};

File: platform/app/controllers/student_controller.py
"""Student CRUD routes — /student/*."""

from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status

from app.models.requests.user_requests import StudentCreateRequest, StudentUpdateRequest
from app.models.responses.login import MessageResponse
from app.models.responses.student_response import StudentResponse
from app.platform.auth.dependencies import require_role
from app.services.user_service import UserService, get_user_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/student", tags=["Students"])


@router.post(
    "",
    summary="Create a student (school_admin only)",
    status_code=status.HTTP_201_CREATED,
    response_model_exclude_none=True,
)
async def create_student(
    body: StudentCreateRequest,
    current_user: dict[str, Any] = Depends(require_role("school_admin")),
    service: UserService = Depends(get_user_service),
) -> StudentResponse:
    if not body.name.strip():
        raise HTTPException(status_code=400, detail="name and phone_number are required")

    school_id = current_user.get("school_id", "")
    tenant_id = current_user.get("tenant_id", "")
    user = await service.create_student(
        name=body.name.strip(),
        phone_number=body.phone_number,
        school_id=school_id,
        tenant_id=tenant_id,
    )
    return StudentResponse.from_domain(user)


@router.get(
    "",
    summary="List students in admin's school",
    status_code=status.HTTP_200_OK,
    response_model_exclude_none=True,
)
async def list_students(
    current_user: dict[str, Any] = Depends(require_role("school_admin", "content_creator", "teacher")),
    service: UserService = Depends(get_user_service),
) -> list[StudentResponse]:
    school_id = current_user.get("school_id", "")
    tenant_id = current_user.get("tenant_id", "")
    if not school_id:
        return []
    students = await service.list_students_for_school(school_id, tenant_id)
    result = [StudentResponse.from_domain(u) for u in students]
    return sorted(result, key=lambda s: s.name)


@router.patch(
    "/{student_id}",
    summary="Update a student (school_admin only)",
    status_code=status.HTTP_200_OK,
    response_model_exclude_none=True,
)
async def update_student(
    student_id: str,
    body: StudentUpdateRequest,
    current_user: dict[str, Any] = Depends(require_role("school_admin")),
    service: UserService = Depends(get_user_service),
) -> StudentResponse:
    if not body.name and not body.phone_number:
        raise HTTPException(status_code=400, detail="name or phone_number is required")

    caller_school = current_user.get("school_id", "")
    updates: dict[str, Any] = {}
    if body.name:
        updates["name"] = body.name.strip()
    if body.phone_number:
        updates["phone"] = body.phone_number

    updated = await service.update_student(student_id, updates, caller_school)
    return StudentResponse.from_domain(updated)


@router.delete(
    "/{student_id}", summary="Delete a student (school_admin only)", status_code=status.HTTP_200_OK
)
async def delete_student(
    student_id: str,
    current_user: dict[str, Any] = Depends(require_role("school_admin")),
    service: UserService = Depends(get_user_service),
) -> MessageResponse:
    caller_school = current_user.get("school_id", "")
    await service.delete_student(student_id, caller_school)
    return MessageResponse(message="Student deleted successfully")
